### PR TITLE
GEODE-9948: Implement LINSERT Command

### PR DESCRIPTION
**Pull request by nonbinaryprogrammer**
Implements the LINSERT command as described by the Redis documentation.

LINSERT key BEFORE|AFTER pivot element

The command inserts an element into a list either BEFORE or AFTER the reference element, which is called the 'pivot'.
It returns the new length of the list if the element is inserted, and -1 if the pivot is not an element of the list.
If the key does not exist, it will not be created, and the command will return -1.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

**Review by jdeppe-pivotal — APPROVED** [review]
Just a minor test addition, otherwise looks good.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.Protocol;
+
+import org.apache.geode.redis.RedisIntegrationTest;
+
+public abstract class AbstractLInsertIntegrationTest implements RedisIntegrationTest {
```
[the remark this answers is not in the document]
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java:36]: I added the test, thank you for thinking of that

**Review by ezoerner — COMMENTED** [review]
Don't we also need to update the docs file? (geode_for_redis.html.md.erb)

File: geode-docs/tools_modules/geode_for_redis.html.md.erb [hunk 2]
```diff
@@ -188,23 +188,23 @@ Could not connect to Redis at 127.0.0.1:6379: Connection refused
...
-| SREM | SSCAN **[3]** | STRLEN | SUBSCRIBE |
-| SUNION | SUNIONSTORE | TTL | TYPE |
-| UNSUBSCRIBE | QUIT | ZADD | ZCARD |
-| ZCOUNT | ZINCRBY | ZINTERSTORE | ZLEXCOUNT |
-| ZPOPMAX | ZPOPMIN | ZRANGE | ZRANGEBYLEX |
-| ZRANGEBYSCORE | ZRANK | ZREM | ZREMRANGEBYLEX |
-| ZREMRANGEBYRANK | ZREMRANGEBYSCORE | ZREVRANGE | ZREVRANGEBYLEX |
-| ZREVRANGEBYSCORE | ZREVRANK | ZSCAN **[3]** | ZSCORE |
-| ZUNIONSTORE ||||
+| LPUSH | LINSERT | MGET | MSET |
```
Comment (ezoerner) [inline geode-docs/tools_modules/geode_for_redis.html.md.erb:191]: Can you keep the list of commands alphabetically sorted?

Comment (nonbinaryprogrammer) [reply geode-docs/tools_modules/geode_for_redis.html.md.erb]: yes, sorry about that. it is fixed now

File: geode-docs/tools_modules/geode_for_redis.html.md.erb [hunk 3]
```diff
@@ -187,24 +187,24 @@ Could not connect to Redis at 127.0.0.1:6379: Connection refused
...
+| SISMEMBER | SMEMBERS | SMOVE | SPOP |
+| SRANDMEMBER | SREM | SSCAN **[3]** | STRLEN |
+| SUBSCRIBE | SUNION | SUNIONSTORE | TTL |
+| TYPE | UNSUBSCRIBE | QUIT | ZADD |
+| ZCARD | ZCOUNT | ZINCRBY | ZINTERSTORE |
+| ZLEXCOUNT | ZPOPMAX | ZPOPMIN | ZRANGE |
+| ZRANGEBYLEX | ZRANGEBYSCORE | ZRANK | ZREM |
+| ZREMRANGEBYLEX | ZREMRANGEBYRANK | ZREMRANGEBYSCORE | ZREVRANGE |
+| ZREVRANGEBYLEX | ZREVRANGEBYSCORE | ZREVRANK | ZSCAN **[3]** |
+| ZSCORE | ZUNIONSTORE ||
```
Comment (ezoerner) [inline geode-docs/tools_modules/geode_for_redis.html.md.erb:207]: The last line should end with 3 vertical bars

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/list/LInsertExecutor.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    if (direction.equalsIgnoreCase("before")) {
+      before = true;
+    } else if (direction.equalsIgnoreCase("after")) {
+      before = false;
+    } else {
+      return RedisResponse.error(RedisConstants.ERROR_SYNTAX);
+    }
+
+    long numEntries = context.listLockedExecute(key, false,
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/list/LInsertExecutor.java:50]: I noticed in another PR that for index we use a 32-bit "int" on lists. I also see that index on this PR's applyInsertByteArrayDelta is an "int".
So why is "numEntries" a long? If the max size of a list is 2G then this can return an int. But if the list can be bigger than 2G then our indexes need to also be "long".

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 5]
```diff
@@ -135,6 +161,27 @@ public int getDSFID() {
...
+          elementList.add(i + 1, elementToInsert);
+          return i + 1;
+        }
+      }
+    }
+
+    return -1;
+  }
+
+  public synchronized void elementInsert(byte[] toInsert, int index) {
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:181]: I could be wrong but this seems like overkill. Just have the single 'int elementInsert' method. Callers are free to ignore its result. If you do keep this you don't need it to be synchronized since it does not do anything that is not thread safe.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 6]
```diff
@@ -135,6 +161,27 @@ public int getDSFID() {
     return REDIS_LIST_ID;
   }
 
+  public synchronized int elementInsert(byte[] elementToInsert, byte[] referenceElement,
+      boolean before) {
+    for (int i = 0; i < elementList.size(); i++) {
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:166]: I think you definitely want to change this to iterate the elements instead of using elementList.get(i). Since elementList is a linked list, every time you call it it has to walk the linked list again to find element 'i'. If you instead just do "for (byte[] element: elementList)" then you will only walk the links once. You will need to also keep you own counter so you know what index you are on but that is easy to do with a local "index" that you increment every time around the for loop,.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 7]
```diff
@@ -135,6 +161,27 @@ public int getDSFID() {
...
+      if (Arrays.equals(elementList.get(i), referenceElement)) {
+        if (before) {
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:168]: Consider something like this:
```
int insertIndex = i;
if (!before) {
  insertIndex++;
}
elementList.add(insertIndex, elementToInsert);
return insertIndex;

```

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/delta/InsertByteArray.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  private final int index;
+
+  public InsertByteArray(byte[] delta, int index) {
+    byteArray = delta;
+    this.index = index;
+  }
+
+  public void serializeTo(DataOutput out) throws IOException {
+    DataSerializer.writeEnum(INSERT_BYTE_ARRAY, out);
+    InternalDataSerializer.writeInteger(index, out);
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/delta/InsertByteArray.java:42]: Change this to DataSerializer.writePrimitiveInt

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/delta/InsertByteArray.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    InternalDataSerializer.writeByteArray(byteArray, out);
+  }
+
+  public static void deserializeFrom(DataInput in, AbstractRedisData redisData) throws IOException {
+    int index = readInteger(in);
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/delta/InsertByteArray.java:47]: Change this to DataSerializer.readPrimitiveInt

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/delta/InsertByteArray.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  public InsertByteArray(byte[] delta, int index) {
+    byteArray = delta;
+    this.index = index;
+  }
+
+  public void serializeTo(DataOutput out) throws IOException {
+    DataSerializer.writeEnum(INSERT_BYTE_ARRAY, out);
+    InternalDataSerializer.writeInteger(index, out);
+    InternalDataSerializer.writeByteArray(byteArray, out);
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/delta/InsertByteArray.java:43]: Change this to DataSerializer. This class should not import InternalDataSerializer.

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java (hunk 5 above)
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:181]: this elementInsert is used when applying the delta. at that point the only information that we have is the index and the new value

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    jedis.lpush(KEY, initialValue);
+
+    assertThat(jedis.linsert(KEY, BEFORE, "nope", insertedValue)).isEqualTo(-1);
+
+    assertThat(jedis.lpop(KEY)).isEqualTo(initialValue);
+    assertThat(jedis.lpop(KEY)).isNull();
+  }
+
+  @Test
+  public void linsert_withInvalidBEFORE_errors() {
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java:75]: Could we also get a test of the behaviour when an invalid BEFORE argument is passed but the key doesn't exist? Just to make sure we're processing things in the same order that Redis does.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+    assertThatThrownBy(() -> jedis.sendCommand(KEY, Protocol.Command.LINSERT, KEY, "LINSERT",
+        "notBefore", initialValue, insertedValue))
+            .isInstanceOf(JedisDataException.class)
+            .hasMessage(
+                String.format("ERR " + RedisConstants.ERROR_WRONG_NUMBER_OF_ARGS, "linsert"));
+  }
+
+  @Test
+  public void linsert_BEFORE_onKeyWithMultipleValues_withValidPivot_insertsValue() {
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java:86]: Could we also get tests for the cases that the list contains duplicate elements, to make sure that we're inserting before or after the right one? For example, for a list `{1, 2, 2, 2, 3}`, calling `LINSERT KEY BEFORE 2 A` should result in `{1, A, 2, 2, 2, 3}` and `LINSERT KEY AFTER 2 A` should result in `{1, 2, A, 2, 2, 3}`

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.Protocol;
+import redis.clients.jedis.exceptions.JedisDataException;
+
+import org.apache.geode.redis.RedisIntegrationTest;
+import org.apache.geode.redis.internal.RedisConstants;
+
+public abstract class AbstractLInsertIntegrationTest implements RedisIntegrationTest {
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java:36]: Could we also add a test of the behaviour when LINSERT is called targeting a key that isn't a list? 

Also a test of the concurrency behaviour using `ConcurrentLoopingThreads`? For this test, we should have a list with some pre-populated elements, and two threads, one calling LINSERT and one modifying the list by adding multiple elements. Then, as the `runWithAction()` section, we should assert that the returned value from LINSERT is one of two values (the expected value if it was called before the elements were inserted, or the expected value if it was called after the elements were inserted) and also that it was inserted at the correct place, using LINDEX. Once the assertions are done, the list should be deleted and then set back to the original value as the final part of the `runWithAction()` to allow the test to continue looping and performing the same actions each time.

**Review by DonalEvans — CHANGES_REQUESTED** [review]
A couple of small suggestions, but really the only changes I'd say were required are using a byte array for comparison rather than a String in the Executor, and adding the described test cases.

File: geode-docs/tools_modules/geode_for_redis.html.md.erb [hunk 14]
```diff
@@ -187,24 +187,24 @@ Could not connect to Redis at 127.0.0.1:6379: Connection refused
...
+| SISMEMBER | SMEMBERS | SMOVE | SPOP |
+| SRANDMEMBER | SREM | SSCAN **[3]** | STRLEN |
+| SUBSCRIBE | SUNION | SUNIONSTORE | TTL |
+| TYPE | UNSUBSCRIBE | QUIT | ZADD |
+| ZCARD | ZCOUNT | ZINCRBY | ZINTERSTORE |
+| ZLEXCOUNT | ZPOPMAX | ZPOPMIN | ZRANGE |
+| ZRANGEBYLEX | ZRANGEBYSCORE | ZRANK | ZREM |
+| ZREMRANGEBYLEX | ZREMRANGEBYRANK | ZREMRANGEBYSCORE | ZREVRANGE |
+| ZREVRANGEBYLEX | ZREVRANGEBYSCORE | ZREVRANK | ZSCAN **[3]** |
+| ZSCORE | ZUNIONSTORE |||
```
[the remark this answers is not in the document]
Comment (nonbinaryprogrammer) [reply geode-docs/tools_modules/geode_for_redis.html.md.erb:207]: yeah that might be nice. we should make a ticket to do something about the editing. seems like everyone dislikes the current method.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 15]
```diff
@@ -135,6 +161,27 @@ public int getDSFID() {
     return REDIS_LIST_ID;
   }
 
+  public int elementInsert(byte[] elementToInsert, byte[] referenceElement,
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:164]: I think this method also needs to be synchronized to protect against a concurrent GII which will serialize this RedisList instance. The GII does not do all the other locking that our redis ops does. So anytime we are making a structural change that is not safe with a concurrent "toData" call we need to add synchronization

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 16]
```diff
@@ -135,6 +161,27 @@ public int getDSFID() {
...
+      boolean before) {
+    int i = 0;
+    for (byte[] element : elementList) {
```
Comment (dschneider-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:167]: I think it would be better to use elementList.listIterator(0) in this code. Then when you find the referenceElement you can just ask the iterator to add elementToInsert without doing another traversal of the linked list which add(i,e) does. Since ListIterator can move either direction you might need to back it up an element before calling add if "before==true". Also if "before==false" and referencedElement is at the end of elementList then I think you will need to call "elementList.add(elementToInsert)" to add it to the end of the list. I think this is worth doing so that you have O(n) instead of 2*O(n) performance.

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  @Test
+  public void shouldDistributeDataAmongCluster_andRetainDataAfterServerCrash() {
+    String key = makeListKeyWithHashtag(1, clusterStartUp.getKeyOnServer("linsert", 1));
+    List<String> elementList = makeElementList(key, INITIAL_LIST_SIZE);
+    lpushPerformAndVerify(key, elementList);
+
+    assertThat(jedis.linsert(key, BEFORE, jedis.lindex(key, 2), insertedValue))
+        .isEqualTo(elementList.size() + 1);
```
Comment (DonalEvans) [inline geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:76]: Could this test also do an LINSERT using AFTER to make sure that the delta propagation for that is working correctly too, just in case?

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+    clusterStartUp.crashVM(1); // kill primary
+
+    assertThat(jedis.llen(key)).isEqualTo(elementList.size() + 1);
+    assertThat(jedis.lindex(key, 2)).isEqualTo(insertedValue);
+    assertThat(jedis.lindex(key, 3)).isEqualTo(elementList.get(INITIAL_LIST_SIZE - 3));
+  }
+
+  @Test
+  public void givenBucketsMoveDuringLInsert_operationsAreNotLost() throws Exception {
```
Comment (DonalEvans) [inline geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:88]: This test is currently only doing 1 single LINSERT for each key. In order to test behaviour during bucket moves effectively, it needs to do repeated LINSERT calls until we've done some reasonable number of bucket moves (about 20 should be good).

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+    keys.add(makeListKeyWithHashtag(2, listHashtags.get(1)));
+    keys.add(makeListKeyWithHashtag(3, listHashtags.get(2)));
+    return keys;
+  }
+
+  private void linsertPerformAndVerify(String key, ListPosition pos, String pivot, String value,
+      AtomicLong runningCount) {
+    long startLength = jedis.llen(key);
+    assertThat(jedis.linsert(key, pos, pivot, value)).isEqualTo(startLength + 1);
+    runningCount.incrementAndGet();
```
Comment (DonalEvans) [inline geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:144]: This running count needs to be incremented immediately before we start calling LINSERT and then either decremented again once we've done some maximum number of repeats or set to 0 if one of the threads doing the LINSERT hits an exception. The way the `lpopPerformAndVerify()` method in `LPopDUnitTest` handles this repeating/stopping behaviour is a pretty good guide.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+            .hasMessage(
+                String.format(WRONG_NUMBER_OF_ARGUMENTS_FOR_COMMAND, "linsert"));
+  }
+
+  @Test
+  public void linsert_onKeyThatIsNotAList_Errors() {
+    jedis.sadd(KEY, initialValue);
+
+    assertThatThrownBy(() -> jedis.linsert(KEY, BEFORE, initialValue, insertedValue))
+        .isInstanceOf(JedisDataException.class)
```
[the remark this answers is not in the document]
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java:80]: uh thought I removed all those

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+              // LINSERT happened first
+              jedisClient -> {
+                assertThat(jedisClient.lindex(KEY, 8))
+                    .as("failure 1")
+                    .isEqualTo(insertedValue);
+                assertThat(jedisClient.lindex(KEY, 0)).isEqualTo("snake");
+              },
+              // LPUSH happened first
+              jedisClient -> assertThat(jedisClient.lindex(KEY, 0))
+                  .as("failure 1").isEqualTo(insertedValue));
```
[the remark this answers is not in the document]
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLInsertIntegrationTest.java:239]: oops I didn't mean to leave these in at all.

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java (hunk 17 above)
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:76]: I don't see the purpose of that. By the time the delta propagation happens the concept of BEFORE and AFTER has been lost

Comment (DonalEvans) [reply geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:76]: True, it's probably overkill. It mostly depends how much we want these tests to be black box tests that are agnostic about implementation details vs. white box tests that test what we know the implementation to be. It's fine to leave this change off.

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java (hunk 18 above)
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:88]: I think this is resolved, let me know if not

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java [hunk 22]
```diff
@@ -98,24 +98,35 @@ public void givenBucketsMoveDuringLInsert_operationsAreNotLost() throws Exceptio
...
 
     Future<Void> future1 = executor.runAsync(task1);
     Future<Void> future2 = executor.runAsync(task2);
     Future<Void> future3 = executor.runAsync(task3);
 
-    for (int i = 0; i < 50 && numInserts.get() > 0; i++) {
+    for (int i = 0; i < 50; i++) {
+      if (i == 20) {
+        continueInserting.set(false);
+      }
```
Comment (DonalEvans) [inline geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:124]: This could be simplified a bit to just have the loop run 20 times, then call `continueInserting.set(false);` afterwards, since we don't care about moving buckets if there are no operations ongoing.

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+    List<String> elementList3 = makeElementList(keys.get(2), INITIAL_LIST_SIZE);
+
+    lpushPerformAndVerify(keys.get(0), elementList1);
+    lpushPerformAndVerify(keys.get(1), elementList2);
+    lpushPerformAndVerify(keys.get(2), elementList3);
+
+    Runnable task1 = () -> {
+      while (continueInserting.get()) {
+        linsertPerformAndVerify(keys.get(0), BEFORE, jedis.lindex(keys.get(0), 2), insertedValue);
+      }
```
Comment (DonalEvans) [inline geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:104]: It might be a little neater to put the while loop inside the `linsertPerformAndVerify()` method.

**Review by DonalEvans — CHANGES_REQUESTED** [review]
The failure in StressNewTest is due to the GII/delta issue, so the test hitting it should be marked `@Ignore` with a reference to the GEODE ticket (GEODE-10108)

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/list/LInsertExecutor.java [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    List<byte[]> commandElements = command.getProcessedCommand();
+
+    String direction = Coder.bytesToString(commandElements.get(2));
+    boolean before;
+    byte[] referenceElement = commandElements.get(3);
+    byte[] elementToInsert = commandElements.get(4);
+
+    if (direction.equalsIgnoreCase("before")) {
+      before = true;
+    } else if (direction.equalsIgnoreCase("after")) {
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/list/LInsertExecutor.java:42]: Rather than creating a String here and doing `equalsIgnoreCase()` it would be more performant to add two new constants to `StringBytesGlossary`:
```
  // LInsertExecutor
  public static final byte[] BEFORE = stringToBytes("BEFORE");
  public static final byte[] AFTER = stringToBytes("AFTER");
```
and compare against them using `Arrays.equals()` here in the executor:
```
    byte[] direction = toUpperCaseBytes(commandElements.get(2));

    if (Arrays.equals(direction, BEFORE)) {
      before = true;
    } else if (Arrays.equals(direction, AFTER)) {
      before = false;
```

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 25]
```diff
@@ -126,6 +127,25 @@ private int getArrayIndex(int listIndex) {
     return listIndex;
   }
 
+  /**
+   * @param elementToInsert element to insert into the set
+   * @param referenceElement element to insert next to
+   * @param before true if inserting before reference element, false if it is after
+   * @param region the region this instance is store in
+   * @param key the name of the list to add to
+   * @return the number of elements in the list after the element is inserted
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:136]: It would be good to include a mention in this comment that this method can also return -1 if the pivot element was not found.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 26]
```diff
@@ -204,6 +229,32 @@ public int getDSFID() {
...
   }
 
+  public synchronized int elementInsert(byte[] elementToInsert, byte[] referenceElement,
+      boolean before) {
+    int i = 0;
+    ListIterator<byte[]> iterator = elementList.listIterator();
+
+    while (iterator.hasNext()) {
+      if (Arrays.equals(iterator.next(), referenceElement)) {
+        if (!before) {
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:239]: Really small nitpick, but could this if statement be inverted, so that we have `if (before) {` at the top? I find it makes it easier to read if we can avoid negation in the conditions.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 27]
```diff
@@ -126,6 +127,25 @@ private int getArrayIndex(int listIndex) {
...
+   */
+  public int linsert(byte[] elementToInsert, byte[] referenceElement, boolean before,
+      Region<RedisKey, RedisData> region, RedisKey key) {
+    int index = elementInsert(elementToInsert, referenceElement, before);
+    if (index == -1) {
+      return index;
+    }
+    storeChanges(region, key, new InsertByteArray(elementToInsert, index));
+
+    return elementList.size();
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:146]: This method should be updated to use versioning in the Delta, similar to what was done for the APPEND command in GEODE-10108.

**Review by DonalEvans — CHANGES_REQUESTED** [review]
Changes are needed to support versioning in the Delta

**Review by dschneider-pivotal — APPROVED** [review]

**Review by DonalEvans — CHANGES_REQUESTED** [review]

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java [hunk 28]
```diff
@@ -136,26 +136,45 @@ public void givenBucketsMoveDuringLInsert_operationsAreNotLost() throws Exceptio
...
+        assertThat(jedis.lindex(key, pivotIndex - 1)).isEqualTo(insertedValue);
+        assertThat(jedis.lindex(key, pivotIndex)).isEqualTo(pivot);
+      } else {
+        assertThat(jedis.lindex(key, pivotIndex + 1)).isEqualTo(insertedValue);
+        assertThat(jedis.lindex(key, pivotIndex)).isEqualTo(pivot);
+      }
+    }
+  }
+
+  private void linsertPerformAndVerify2(String key, ListPosition pos, int pivotIndex,
```
Comment (DonalEvans) [inline geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:160]: I think there are accidentally two identical copies of this method now. This one should be removed.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

File: geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java (hunk 28 above)
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LInsertDUnitTest.java:160]: yeah, I saw that. I'm looking at the stressNewTest failures too

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 29]
```diff
@@ -127,6 +128,32 @@ private int getArrayIndex(int listIndex) {
     return listIndex;
   }
 
+  /**
+   * @param elementToInsert element to insert into the set
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:132]: This should be "element to insert into the list"

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 30]
```diff
@@ -229,6 +261,32 @@ public int getDSFID() {
     return REDIS_LIST_ID;
   }
 
+  public synchronized int elementInsert(byte[] elementToInsert, byte[] referenceElement,
+      boolean before) {
+    int i = 0;
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:266]: Rather than using a separate counter here, you could instead use the `nextIndex()` or `previousIndex()` methods in `ListIterator` to return the appropriate index. Really the only reason to create a `ListIterator` rather than a plain `Iterator` here is to have access to those methods.

**Review by DonalEvans — APPROVED** [review]
A couple of small comments, but nothing that should stop this being merged.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java (hunk 30 above)
Comment (nonbinaryprogrammer) [reply geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:291]: for some reason this didn't work. I looked at it for awhile but don't see why it wouldn't. I left the code as is.
